drv/sidecar-front-io/src/transceivers.rs

```diff
@@ -77,6 +90,144 @@ impl FpgaPortMasks {
+
+    pub fn get(&self) -> u32 {
+        self.0
+    }
+    pub fn set(&mut self, index: LogicalPort) {
+        self.0 |= index.as_mask().0
+    }
+    pub fn clear(&mut self, index: LogicalPort) {
+        self.0 &= !index.as_mask().0
+    }
+    pub fn merge(&mut self, other: LogicalPortMask) {
+        self.0 |= other.0
+    }
+    pub fn count(&self) -> usize {
+        self.0.count_ones() as _
+    }
+    pub fn is_set(&self, index: LogicalPort) -> bool {
+        self.0 & index.as_mask().0 != 0
+    }
+    pub fn is_empty(&self) -> bool {
+        self.0 == 0
+    }
+    pub fn to_indices(&self) -> impl Iterator<Item = u8> + '_ {
+        (0..32).filter(|i| self.is_set(LogicalPort(*i)))
+    }
+}
+
+impl From<ModuleId> for LogicalPortMask {
+    fn from(v: ModuleId) -> Self {
+        Self(v.0 as u32)
+    }
+}
+
+impl From<LogicalPortMask> for ModuleId {
+    fn from(v: LogicalPortMask) -> Self {
+        Self(v.0 as u64)
+    }
+}
+
+// It is convenient to have the ergonomics for a LogicalPortMask resemble the
+// bitwise mask that it represents, so we implement some bitwise operations.
+impl core::ops::BitOr for LogicalPortMask {
+    type Output = Self;
+
+    fn bitor(self, rhs: Self) -> Self {
+        LogicalPortMask(self.0 | rhs.0)
+    }
+}
+
+impl core::ops::BitOrAssign for LogicalPortMask {
+    fn bitor_assign(&mut self, rhs: Self) {
+        *self = *self | rhs
+    }
+}
+
+impl core::ops::BitAnd for LogicalPortMask {
+    type Output = Self;
+
+    fn bitand(self, rhs: Self) -> Self {
+        LogicalPortMask(self.0 & rhs.0)
+    }
+}
+
+impl core::ops::BitAndAssign for LogicalPortMask {
+    fn bitand_assign(&mut self, rhs: Self) {
+        *self = *self & rhs
+    }
+}
+
+impl core::ops::Not for LogicalPortMask {
+    type Output = Self;
+
+    fn not(self) -> Self {
+        LogicalPortMask(!self.0)
+    }
+}
+
+// Maps physical port `mask` to a logical port mask
+impl From<FpgaPortMasks> for LogicalPortMask {
+    fn from(mask: FpgaPortMasks) -> LogicalPortMask {
+        let mut logical_mask: u32 = 0;
+        for port in 0..NUM_PORTS {
+            let logical_port = PORT_MAP[port as usize];
+            let bits = match logical_port.controller {
+                FpgaController::Left => mask.left,
+                FpgaController::Right => mask.right,
+            };
+            let port_mask = logical_port.port.as_mask();
+            if bits.is_set(logical_port.port)
+                && port_mask.is_set(logical_port.port)
+            {
+                logical_mask |= 1 << port;
+            }
+        }
+
+        LogicalPortMask(logical_mask)
+    }
+}
+
+// Maps logical port `mask` to physical FPGA locations
+impl From<LogicalPortMask> for FpgaPortMasks {
+    fn from(mask: LogicalPortMask) -> FpgaPortMasks {
+        let mut fpga_port_masks = FpgaPortMasks::default();
+        for (i, port_loc) in PORT_MAP.iter().enumerate() {
+            let port_mask: u32 = 1 << i;
+            if (mask.0 & port_mask) != 0 {
```

Comment:
Continuing my campaign for Stronger Types, how about
```suggestion
            let this_port_mask = LogicalPort(i as u8);
            if mask.is_set(this_port_mask) {
```

Reply:
Done locally.